### PR TITLE
Adding Reverse op

### PR DESCRIPTION
**Pull request by pk-g**

### CODE REVIEW

File: onnx/defs/tensor/defs.cc [hunk 1]
```diff
@@ -622,6 +622,75 @@ ONNX_OPERATOR_SET_SCHEMA(
...
+    OpSchema()
+        .SetDoc(Reverse_ver9_doc)
+        .Attr(
+            "axes",
+            "A list of integers indicating the axes to reverse. "
+            "Negative values mean counting dimensions from the back."
+            "By default, reverse over all axes of input dimensions",
+            AttributeProto::INTS,
+            OPTIONAL)
+        .Input(0, "inpu6", "Tensor of rank r >= 1.", "T")
```
Comment (gramalingam) [inline onnx/defs/tensor/defs.cc]: inpu6 => input

File: onnx/defs/operator_sets.h [hunk 2]
```diff
@@ -497,6 +497,7 @@ class ONNX_OPERATOR_SET_SCHEMA_CLASS_NAME(Onnx, 9, Where);
 class ONNX_OPERATOR_SET_SCHEMA_CLASS_NAME(Onnx, 9, Cast);
 class ONNX_OPERATOR_SET_SCHEMA_CLASS_NAME(Onnx, 9, NonZero);
 class ONNX_OPERATOR_SET_SCHEMA_CLASS_NAME(Onnx, 9, TfIdfVectorizer);
+class ONNX_OPERATOR_SET_SCHEMA_CLASS_NAME(Onnx, 9, Reverse);
```
Comment (linkerzhang) [inline onnx/defs/operator_sets.h]: the operator set should be 10. 

9 was the one released in ONNX 1.4.

Comment (houseroad) [reply onnx/defs/operator_sets.h]: Please use 10 instead.

**Comment by pk-g** [comment]
Thanks for the review, updated the PR with addressed feedback.

File: onnx/backend/test/case/node/reverse.py [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+        input = np.arange(6.0).reshape(2, 3)
+
+        node = onnx.helper.make_node(
+            'Reverse',
+            inputs=['input'],
+            outputs=['output'],
+            axes=[0]
+        )
+
+        output = np.flip(input)
```
Comment (gramalingam) [inline onnx/backend/test/case/node/reverse.py]: Is this correct? Shouldn't the axes=0 be passed to np.flip?

Comment (pk-g) [reply onnx/backend/test/case/node/reverse.py]: the default is to reverse on all axes.

**Comment by pengwa** [comment]
shall we enhance this op to support reverse sequences of different lengths? in tf, it is called tf.reversesequence().

**Comment by pk-g** [comment]
@pengwa  If you are referring to the case where sequential data of different lengths has been _padded_, it may be cleaner to have a separate op to cover such scenario.

**Comment by pk-g** [comment]
@linkerzhang can you please review the updated PR so we can move ahead with this op? thanks.

File: docs/Changelog.md [hunk 4]
```diff
@@ -9451,6 +9451,83 @@ This version of the operator has been available since version 9 of the default O
...
+
+#### Version
+
+This version of the operator has been available since version 10 of the default ONNX operator set.
+
+#### Attributes
+
+<dl>
+<dt><tt>axes</tt> : list of ints</dt>
+<dd>A list of integers indicating the axes to reverse. Negative values mean counting dimensions from the back.By default, reverse over all axes of input dimensions</dd>
```
Comment (hariharans29) [inline docs/Changelog.md]: Sorry for this clarification so late on - but I assume this operator doesn't support "implicit" copying of Tensors by specifying duplicates in axes ?  For example, would axes of [0,0] mean no reversing but force the tensor's buffer to be copied ? Or simply, don't allow duplicates in axes at all - just like the way tf operates - https://www.tensorflow.org/api_docs/python/tf/reverse ?

Comment (pk-g) [reply docs/Changelog.md]: Great question, this is more in line with the latter, no duplicate is expected here.

**Comment by linkerzhang** [comment]
@pk-g can you update the PR by fixing the conflicts?

@houseroad any more comments on this op please?

Thank you!

**Comment by houseroad** [comment]
I will do the review later this week, but please don't rush.

Will revert it here: https://github.com/onnx/onnx/pull/1882

Please resubmit the changes.

**Comment by fdwr** [comment]
> Will revert it here: #1882
> 
> Please resubmit the changes.

Was this op reverted and then simply forgotten to re-add? It's found in the following libraries...

- PyTorch [flip](https://pytorch.org/docs/stable/generated/torch.flip.html)
TensorFlow [reverse](https://www.tensorflow.org/api_docs/python/tf/reverse)
- CoreML [tensor_transformation.reverse](https://apple.github.io/coremltools/source/coremltools.converters.mil.mil.ops.defs.html#coremltools.converters.mil.mil.ops.defs.iOS17.tensor_transformation.reverse)
- TFLite [tfl.reverse_v2](https://www.tensorflow.org/mlir/tfl_ops#tflreverse_v2_tflreversev2op)
- StableHLO reverse
- tosa.reverse (mlir::tosa::ReverseOp)
- ANEURALNETWORKS_REVERSE
- PyTorch prims rev

...and we're looking to [add it to WebNN](https://github.com/webmachinelearning/webnn/issues/773) for a voice to text model. It *can* be achieved in ONNX via the overloaded interpretation of setting ONNX [Slice](https://onnx.ai/onnx/operators/onnx__Slice.html)'s to -1, but loses semantics and is more awkward (needing to specify `starts: [2147483647, 2147483647], ends: [-2147483648, -2147483648], steps: [-1,-1], axes: [0,1]` rather than just `axes: [0,1]`).

**Pull request merged**